ksqldb-tools/src/main/java/io/confluent/ksql/tools/migrations/commands/ApplyMigrationCommand.java

```diff
@@ -313,26 +313,74 @@ private void executeCommands(
       final Clock clock,
       final String previous
   ) {
-    final Map<String, Object> properties = new HashMap<>();
     final List<String> commands = CommandParser.splitSql(migrationFileContent);
 
+    executeCommands(
+        commands, ksqlClient, config, executionStart, migration, clock, previous, true);
+    executeCommands(
+        commands, ksqlClient, config, executionStart, migration, clock, previous, false);
+  }
+
+  private void executeCommands(
+      final List<String> commands,
+      final Client ksqlClient,
+      final MigrationConfig config,
+      final String executionStart,
+      final MigrationFile migration,
+      final Clock clock,
+      final String previous,
+      final boolean verify
+  ) {
+    cleanUpJavaClientVariables(ksqlClient);
+    final Map<String, Object> properties = new HashMap<>();
     for (final String command : commands) {
       try {
         final Map<String, String> variables = ksqlClient.getVariables().entrySet()
             .stream().collect(Collectors.toMap(e -> e.getKey(), e -> e.getValue().toString()));
         executeCommand(
-            CommandParser.transformToSqlCommand(command, variables), ksqlClient, properties);
-      } catch (InterruptedException | ExecutionException e) {
+            CommandParser.transformToSqlCommand(command, variables),
+            ksqlClient,
+            properties,
+            verify
+        );
+      } catch (InterruptedException | ExecutionException | MigrationException e) {
+        final String action = verify ? "parse" : "execute";
         final String errorMsg = String.format(
-            "Failed to execute sql: %s. Error: %s", command, e.getMessage());
+            "Failed to %s sql: %s. Error: %s", action, command, e.getMessage());
         updateState(config, ksqlClient, MigrationState.ERROR,
             executionStart, migration, clock, previous, Optional.of(errorMsg));
         throw new MigrationException(errorMsg);
       }
     }
   }
 
+  private void cleanUpJavaClientVariables(final Client ksqlClient) {
+    ksqlClient.getVariables().forEach((k, v) -> ksqlClient.undefine(k));
+  }
+
   private void executeCommand(
+      final SqlCommand command,
+      final Client ksqlClient,
+      final Map<String, Object> properties,
+      final boolean defineUndefineOnly
```

Comment:
As above, we need comments explaining this, both what it does and why it's useful. (Up to you where the comments go, no need to duplicate.)